docs/decisions/0017-generalized-containers.rst

```diff
@@ -0,0 +1,70 @@
+17. Modeling Containers as a Generalized Capability for Holding Content
+========================================================================
+
+Context
+-------
+
+This ADR proposes a model for containers that can hold different types of content and can be used to model other content types with similar behavior, such as units, subsections, sections, or courses. The model defines containers' core structure and purpose, the types of containers, content constraints, container members, version control, publishing, and pruning.
+
+Decisions
+---------
+
+1. Core Structure and Purpose of Containers
+===========================================
+
+- A container is designed as a generalized capability to hold different types of content.
+- A container is a publishable content type that holds other content types through a parent-child relationship.
+- A container application will offer shared definitions for use by other container types.
+
+2. Container Types and Content Constraints
+==========================================
+
+- A container marks any PublishableEntity, such as sections, subsections, units, or any other custom content type, as a type that can hold other content.
+- Containers can be nested within other containers, allowing for complex content structures.
+- Containers might be of different types, with each type potentially having different restrictions on the type of content it can hold but that will not be enforced by containers.
+- Content restrictions for containers are implemented at the app layer, allowing specific container types, like Unit, to limit their members to particular content types (e.g., only Components).
+- The course hierarchy Course > Section > Subsection > Unit will be implemented as relationships between containers, with each level acting as a container that holds other content. The hierarchy will be enforced by the content restrictions of each particular container but allowed to be overridden to support `0002-content-flexibility.rst`_.
+- Containers will follow extensibility principles in `0003-content-extensibility.rst`_ for creating new container types or subtypes.
+
+3. Container Members and Relationships
+=======================================
+
+- The members of a container can be any type of publishable content.
+- Members within a container are maintained in a specific order as an ordered list.
+- Containers represent their content hierarchy through a structure that defines parent-child relationships between the container and its members.
+- The structure defining these relationships is anonymous, so it can only be referenced through the container.
+- Containers can hold both static and dynamically generated content, including user-specific variations.
+- Each container holds different states of its members (user-defined, initial, and frozen final state) to support rollback operations.
```

Comment:
I think an expanded description of these three states would be helpful. From this I'm not totally clear on what's the difference between initial and user-defined. And why "initial" is immutable but "frozen" is not (?) even though it sounds like it should be from the name.

Reply:
For the expanded descriptions, I created a separate section for "Container States," grouping all the state management decisions instead: https://github.com/openedx/openedx-learning/pull/251/commits/af7dce4eb392ffa8967f7c9333ffb80c4f154489. I think now each definition is clear according to what was proposed in the model sketches.

Reply:
> why "initial" is immutable but "frozen" is not (?) even though it sounds like it should be from the name

I understand frozen lists are "this is how the container looked (a screenshot?) when a new version was created". So, if there's a unit U0 with components `defined_list = P0latest, P1latest, P2latest` all pointing to their latest versions at the time of the next container creation where `frozen_list` would be `None` and `initial_list = P0, P1, P2` where P0, P1, and P2 are the pinned versions of `P0latest `and so on at moment of creation. When U1 is created, `defined_list` looks like `defined_list = P0latest, P1latest, P2latest` so frozen_list would be `frozen_list = P0', P1'', P2'` (the pinned versions of P0latest and so on). New versions for P0,...,1 are created, we'd need a way of knowing what versions were in U1 in case we want to go back to it, so we use frozen lists.

Frozen lists are useful when there are unpinned versions of members, since when they're all pinned, then `defined_list = initial_list = frozen_list`.

IThese in-line comments Dave left in #240 might explain this better: https://github.com/openedx/openedx-learning/pull/254/files#diff-6f2c589dc4ba5960e91d39f6488eb5e2e2e63ddaff63a75909091c760b877802R112-R154

Reply:
Thanks for expanding the descriptions, but I'm still confused :/

The code seems to call these "lists" not "states" - it would be good to pick a consistent term.

> The user-defined state of a container is the state that the author has defined for the version of the container

I don't think we should call this "user-defined" when it's actually "author-defined." To me, "user-defined" sounds more like "the version of this container that a specific user will see (after we've accounted for groups, A/B testing, randomized content, exam permissions, etc.). But "author-defined" is clear.

> - The initial state of a container is the state of the container when it was first created.

>    `# inital_list is an EntityList where all the versions are pinned, to show`
>    `# what the exact versions of the children were at the time that the`
>    `# Container was created.`

Is initial list really the list when the **container** was first created? (which I assume would usually be an empty list). I _think_ this should say "the initial list is a copy of the author-defined list that has all versions pinned as they were at the time the **ContainerEntityVersion** was created." Because if this is really something immutable from when the _container_ was created, not something for when each _version_ is created, then it should be on the container model, not redundantly specified for every ContainerEntityVersion.

> The frozen final state of a container is the state of the container at the time when a new version is created.

I think this needs an example: "While this ContainerEntityVersion is the current draft, this will be None which means that unpinned entities in the list should be showing their latest version. However, once this is published or an even newer draft is created, this frozen list should be saved to reflect the history of the list at that exact point in time when the version was finalized."

---------

I also don't really understand the need for an `initial_list` at all. Isn't it the same as "the frozen_list of the previous version"? And whether that's the case or not, when/why do we need to use it (initial_list)? I don't see any explanation in https://github.com/openedx/openedx-learning/issues/240 .

Reply:
Oh, now I see that in the ADR itself I called it container states in some places and in others lists. I updated this also considering the author-defined suggestion:  https://github.com/openedx/openedx-learning/pull/251/commits/506d9cdc3a3dc53df3fdd2d33e8acfac95dae550

Reply:
Oh, now I see that in the ADR itself I called it container states in some places and in others lists. I updated this also considering the author-defined suggestion:  https://github.com/openedx/openedx-learning/pull/251/commits/506d9cdc3a3dc53df3fdd2d33e8acfac95dae550

Reply:
> Is initial list really the list when the container was first created? (which I assume would usually be an empty list). 

No. It's when the next container version is created, not the container entity itself. 

> I think this should say "the initial list is a copy of the author-defined list that has all versions pinned as they were at the time the ContainerEntityVersion was created." 

Yes, that's a better description of what's actually happening here. However, I want to keep this as high-level as possible, so I won't mention container entities or other model-specific definitions.

> Because if this is really something immutable from when the container was created, not something for when each version is created, then it should be on the container model, not redundantly specified for every ContainerEntityVersion.

Exactly, these lists are actually related to each container version, not the container entity. I'll make that clear.

Reply:
Oh, now I see that in the ADR itself I called it container states in some places and in others lists. I updated this also considering the author-defined suggestion. Thanks for the suggestions!

---

> Is initial list really the list when the container was first created? (which I assume would usually be an empty list). 

No. It's when the next container version is created, not the container entity itself. 

> I think this should say "the initial list is a copy of the author-defined list that has all versions pinned as they were at the time the ContainerEntityVersion was created." 

Yes, that's a better description of what's actually happening here. However, I want to keep this as high-level as possible, so I won't mention container entities or other model-specific definitions unless you consider them necessary.

> Because if this is really something immutable from when the container was created, not something for when each version is created, then it should be on the container model, not redundantly specified for every ContainerEntityVersion.

Exactly, these lists are actually related to each container version, not the container entity. I'll make that clear.

---

> I think this needs an example: "While this ContainerEntityVersion is the current draft, this will be None which means that unpinned entities in the list should be showing their latest version. However, once this is published or an even newer draft is created, this frozen list should be saved to reflect the history of the list at that exact point in time when the version was finalized."

In "...unpinned entities in the list should be showing their lates version"  do you refer to the author-defined list, right?

---

> I also don't really understand the need for an initial_list at all. Isn't it the same as "the frozen_list of the previous version"? And whether that's the case or not, when/why do we need to use it (initial_list)? I don't see any explanation in https://github.com/openedx/openedx-learning/issues/240 .

I think the idea of an initial state came from this comment: https://github.com/openedx/openedx-learning/issues/38#issuecomment-2395589146 -- still, it doesn't say the end purpose. But as far as I understand, it's included for convenience according to: 

    # inital_list is an EntityList where all the versions are pinned, to show
    # what the exact versions of the children were at the time that the
    # Container was created. We could technically derive this, but it would be
    # awkward to query.
    #
    # If the Container was defined so that all references were pinned, then this
    # can point to the exact same EntityList as defined_list.

Reply:
Yeah, I saw that, but I still don't understand the use case. I also can't tell if it's any different than "the frozen_list of the previous version", or if those are the same thing.

@ormsbee maybe you can clarify?